dashdotdb/controllers/tokens.py

```diff
@@ -0,0 +1,42 @@
+from datetime import datetime
+from uuid import uuid4
+
+from dashdotdb.models.dashdotdb import db
+from dashdotdb.models.dashdotdb import Token
+from dashdotdb.models.dashdotdb import LatestTokens
+
+
+def close_token(uuid, data_type):
+    db_token = db.session.query(Token) \
+        .filter(Token.uuid == uuid, Token.data_type == data_type).first()
+    if db_token is None:
+        return 'token not found', 404
+    with db.session.begin():
+        db_token.is_open = False
+        latest_token = db.session.query(LatestTokens) \
+            .filter(Token.data_type == data_type,
+                    Token.id == LatestTokens.token_id).first()
+        if latest_token is None:
+            db.session.add(LatestTokens(token_id=db_token.id))
+        else:
+            # Only update the latest token if the creation timestamp is newer than
+            # the current latest token creation timestamp
+            latest_token_data = db.session.query(
+                Token).filter(Token.id == latest_token.token_id).first()
+            if db_token.timestamp > latest_token_data.timestamp:
+                latest_token.token_id = db_token.id
+    return 'token closed'
+
+
+def generate_token(data_type):
+    # generate a unique id
+    uuid = str(uuid4())
+    db_token = db.session.query(Token) \
+        .filter(Token.uuid == uuid).first()
+    if db_token is None:
+        db.session.add(Token(uuid=uuid, data_type=data_type,
+                             timestamp=datetime.now(),
+                             is_open=True))
+        db.session.commit()
+        return uuid
+    return 'token collision', 500
```

Comment:
if we know this is a condition we might reach, can we handle it instead of returning 500? Maybe retry a couple of times? I'd not care about collisions at all, but I understand the concern :)

Reply:
If the uuid library is functioning correctly this should never happen.  Do we want to try to account for that library returning a re-used UUID?

Reply:
You're already accounting for that. That's why I mentioning making it safer, but to be honest, I'd just trust it returns the right thing and drop the:
```
    db_token = db.session.query(Token) \
        .filter(Token.uuid == uuid).first()
    if db_token is None:
```
This will even make things a bit faster and a bit less db intensive.

Reply:
I suppose that's true.  we might as well trust the library all the way.

Reply:
Done